### PR TITLE
[INFRA-35557] ReactiveAerospikeTemplate implementation

### PR DESCRIPTION
**Pull request by igor-ermolenko**
Implemented save-related methods

### CODE REVIEW

**Comment by tdanylchuk** [comment]
cas tests?

**Comment by igor-ermolenko** [comment]
> cas tests?
Added ```testRejectSaveWithOldVersion```

File: pom.xml [hunk 1]
```diff
@@ -174,10 +186,21 @@
             <artifactId>aerospike-client</artifactId>
         </dependency>
 
+        <dependency>
+            <groupId>com.aerospike</groupId>
+            <artifactId>aerospike-reactor-client</artifactId>
```
Comment (Aloren) [inline pom.xml]: this should be optional dependency

**Comment by Aloren** [comment]
develop in this fork is out of date. check out master from this repo: https://github.com/aerospike/spring-data-aerospike

File: pom.xml [hunk 2]
```diff
@@ -174,10 +186,21 @@
...
+        </dependency>
+
         <dependency>
             <groupId>com.aerospike</groupId>
             <artifactId>aerospike-helper-java</artifactId>
         </dependency>
+
+        <dependency>
+            <groupId>io.projectreactor</groupId>
+            <artifactId>reactor-core</artifactId>
```
Comment (Aloren) [inline pom.xml]: this should be optional dependency

File: src/main/java/org/springframework/data/aerospike/config/AbstractAerospikeDataConfiguration.java [hunk 3]
```diff
@@ -37,6 +40,15 @@ public AerospikeTemplate aerospikeTemplate(AerospikeClient aerospikeClient,
         return new AerospikeTemplate(aerospikeClient, nameSpace(), mappingAerospikeConverter, aerospikeMappingContext, aerospikeExceptionTranslator);
     }
 
+    @Bean(name = "reactiveAerospikeTemplate")
+    public ReactiveAerospikeTemplate reactiveAerospikeTemplate(AerospikeClient aerospikeClient,
```
Comment (Aloren) [inline src/main/java/org/springframework/data/aerospike/config/AbstractAerospikeDataConfiguration.java]: please create separate `AbstractReactiveAerospikeDataConfiguration`. see an example here: https://github.com/spring-projects/spring-data-couchbase/blob/master/src/main/java/org/springframework/data/couchbase/config/AbstractReactiveCouchbaseDataConfiguration.java

File: src/main/java/org/springframework/data/aerospike/core/ReactiveAerospikeTemplate.java [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    private final AerospikeReactorClient reactorClient;
+
+    public ReactiveAerospikeTemplate(AerospikeClient client,
+                                     String namespace,
+                                     MappingAerospikeConverter converter,
+                                     AerospikeMappingContext mappingContext,
+                                     AerospikeExceptionTranslator exceptionTranslator,
+                                     AerospikeReactorClient reactorClient) {
+        super(client, namespace, converter, mappingContext, exceptionTranslator);
+        this.reactorClient = reactorClient;
```
Comment (Aloren) [inline src/main/java/org/springframework/data/aerospike/core/ReactiveAerospikeTemplate.java]: please add
Assert.notNull(reactorClient, "Aerospike reactor client must not be null!");

File: src/main/java/org/springframework/data/aerospike/core/ReactiveAerospikeTemplate.java [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+                    .map(docKey -> document);
+        } catch (AerospikeException e) {
+            throw translateIfPossible(e);
+        }
+    }
+
+    private <T> Mono<T> doPersist(T document, WritePolicy policy) {
+        try {
+            AerospikeWriteData data = writeData(document);
+            return reactorClient.put(policy, data.getKey(), data.getBinsAsArray())
```
Comment (Aloren) [inline src/main/java/org/springframework/data/aerospike/core/ReactiveAerospikeTemplate.java]: try-catch won't work in any of these methods. you need to use onError method for exception translation

File: src/test/java/org/springframework/data/aerospike/core/ReactiveAerospikeTemplateIntegrationTests.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+import static com.aerospike.client.policy.RecordExistsAction.UPDATE_ONLY;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Integration tests for {@link AerospikeTemplate}.
+ *
+ * @author Igor Ermolenko
+ */
+public class ReactiveAerospikeTemplateIntegrationTests extends BaseIntegrationTests {
```
Comment (Aloren) [inline src/test/java/org/springframework/data/aerospike/core/ReactiveAerospikeTemplateIntegrationTests.java]: please check the tests present in https://github.com/aerospike/spring-data-aerospike/blob/master/src/test/java/org/springframework/data/aerospike/core/AerospikeTemplateTests.java
you need to add at least the ones that are not related to different data types

File: src/main/java/org/springframework/data/aerospike/core/ReactiveAerospikeTemplate.java [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+        }
+    }
+
+    private <T> Mono<T> doPersistWithCas(T document, AerospikePersistentEntity<?> entity) {
+        try {
+            AerospikeWriteData data = writeData(document);
+            ConvertingPropertyAccessor accessor = getPropertyAccessor(entity, document);
+            WritePolicy policy = getCasAwareWritePolicy(data, entity, accessor);
+            Operation[] operations = OperationUtils.operations(data.getBinsAsArray(), Operation::put, Operation.getHeader());
+            return reactorClient.operate(policy, data.getKey(), operations)
```
Comment (kptfh) [inline src/main/java/org/springframework/data/aerospike/core/ReactiveAerospikeTemplate.java]: I'd rather use reactorClient.put instead of operate(). Operate returns bins data that is ignored in this case.

File: pom.xml (hunk 1 above)
Comment (igor-ermolenko) [reply pom.xml]: done

File: pom.xml (hunk 2 above)
Comment (igor-ermolenko) [reply pom.xml]: done

File: src/main/java/org/springframework/data/aerospike/core/ReactiveAerospikeTemplate.java (hunk 5 above)
Comment (igor-ermolenko) [reply src/main/java/org/springframework/data/aerospike/core/ReactiveAerospikeTemplate.java]: Fixed

File: src/main/java/org/springframework/data/aerospike/core/ReactiveAerospikeTemplate.java (hunk 4 above)
Comment (igor-ermolenko) [reply src/main/java/org/springframework/data/aerospike/core/ReactiveAerospikeTemplate.java]: done

File: src/test/java/org/springframework/data/aerospike/core/ReactiveAerospikeTemplateIntegrationTests.java (hunk 6 above)
Comment (igor-ermolenko) [reply src/test/java/org/springframework/data/aerospike/core/ReactiveAerospikeTemplateIntegrationTests.java]: In this PR added tests only for implemented methods (save, persist, insertAll, insert, update).
Other tests will be added during other methods implementation.
And at the end of all will be added all absent tests.
Is this ok?

File: src/main/java/org/springframework/data/aerospike/config/AbstractAerospikeDataConfiguration.java (hunk 3 above)
Comment (igor-ermolenko) [reply src/main/java/org/springframework/data/aerospike/config/AbstractAerospikeDataConfiguration.java]: I've created AbstractReactiveAerospikeDataConfiguration with React related stuff.
All other infrastructure will be implemented in next PRs together with ReactiveAerospikeRepository implementation.

File: src/main/java/org/springframework/data/aerospike/core/BaseAerospikeTemplate.java [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    WritePolicy getCasAwareWritePolicy(AerospikeWriteData data, AerospikePersistentEntity<?> entity,
+                                               ConvertingPropertyAccessor accessor) {
+        WritePolicyBuilder builder = WritePolicyBuilder.builder(this.client.writePolicyDefault)
+                .sendKey(true)
+                .generationPolicy(GenerationPolicy.EXPECT_GEN_EQUAL)
+                .expiration(data.getExpiration());
+
+        Integer version = accessor.getProperty(entity.getVersionProperty(), Integer.class);
```
Comment (VolodymyrShpynta) [inline src/main/java/org/springframework/data/aerospike/core/BaseAerospikeTemplate.java]: There could be ClassCastException in the case when the entity @Version field is a long type.

Comment (Aloren) [reply src/main/java/org/springframework/data/aerospike/core/BaseAerospikeTemplate.java]: no, this is safe conversion
